### ISSUE TITLE
`undefined` type from optional chaining doesn't produce correct generic inference candidate from union

### ISSUE THREAD

**Issue by RyanCavanaugh** [microsoft/TypeScript#34579]
**TypeScript Version:**  10/18/2019 nightly

**Search Terms:** optional chaining generic inference undefined

**Code**

```ts
declare function unbox<T>(box: { value: T | undefined }): T;
declare const su: string | undefined;

const b1 = { value: su?.length };
// Incorrect error - T inferred to be number | undefined
const v1: number = unbox(b1);

// No error
const b2 = { value: su?.length as number | undefined };
const v2: number = unbox(b2);

// No error
const b3: { value: number | undefined } = { value: su?.length  };
const v3: number = unbox(b3);
```

**Expected behavior:** In all cases, `T` should be inferred to be `number`

**Actual behavior:** When the `undefined` part of the union of the type of `value` originates in an optional chaining, it doesn't infer correctly

**Playground Link:** http://www.typescriptlang.org/play/?ts=Nightly&ssl=1&ssc=1&pln=15&pc=1#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXzQCMcAPAHgBUA+ACmJIC54BveANygmRCYvgB8CqUIiyoQweAF8AlLwDcAWABQoSLARg8AZwzxtyJrphiA5gKEixEpcpVbUu+IQCM8ALwt2nbkeQB+ADoIEFRTDAALaVsAehj4AEl0HBg4THgQVJT4AFp4PjFETLhJDBxnBFRkAFtCTIs0K3Fgex09NhcmKtr6zyJSOhcZWxU4+AA5cuKU1sc9QgAmDy8OLh59AODQ8KiobXhuupgG4RBRZujZpzYFrpqj5f6SOgXhlVH4yYysmCv5gGYmKxVr4Dvd6oJGmdrJIpMtgT51gYgiEwpF4JdlA5roCwT1jn1UPQ6P83sogA (nightly link)

**Related Issues:** Nope

**Comment by RyanCavanaugh** [comment]
cc @rbuckton who was looking at this with us

**Comment by JannesMeyer** [comment]
I think I've come across the same issue. This is my code:

```ts
let values = { a: 1 } as { a: number } | undefined;

let a = values?.a;
assertDefined(a); // This does not narrow the type to number
console.log(a.toString()); // Error: Object is possibly undefined

function assertDefined<T>(x: T | undefined | null): asserts x is T {
    if (x == null) {
        throw new Error('Expected a value');
    }
}
```

[Playground link](https://www.typescriptlang.org/play/?ts=3.7-Beta&ssl=1&ssc=1&pln=24&pc=2#code/DYUwLgBAbghsCuIDOEC8EDeEYC4IEYIBfbFLXCAO3gFsAjEAJ2IgB8J5KATEAMwEtKILgG4AUGNCQYaaHERIA-ADoY4mEiRMwAET6DhAChgBKERAD0FiABUAFvxRcA9sirPIlGI0bOA7hBgdiCBAJ4ADiFgzlS0DIxiAMbOlEjOoMrAzgDmxsrRAMpgjIK5JmYSvJyJYPwppFqMuvpCXAA8NgB8hgAeeDZsHNwtwoPUwMAmeBqNYCg9EI62mGIQa4u8EL1o6OOTK+uHgXa+AUIBAKI+zoyGAOQXPZE1ozKwCCB3FYdEYkRAA)

**Comment by rosslovas** [comment]
This is in the 3.7.2 milestone but the fix didn't actually make it in ([playground link](https://www.typescriptlang.org/play/#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXzQCMcAPAHgBUA+ACmJIC54BveANygmRCYvgB8CqUIiyoQweAF8AlLwDcAWABQoSLARg8AZwzxtyJrphiA5gKEixEpcpVbUu+IQCM8ALwt2nbkeQB+ADoIEFRTDAALaVsAehj4AEl0HBg4THgQVJT4AFp4PjFETLhJDBxnBFRkAFtCTIs0K3Fgex09NhcmKtr6zyJSOhcZWxU4+AA5cuKU1sc9QgAmDy8OLh59AODQ8KiobXhuupgG4RBRZujZpzYFrpqj5f6SOgXhlVH4yYysmCv5gGYmKxVr4Dvd6oJGmdrJIpMtgT51gYgiEwpF4JdlA5roCwT1jn1UPQ6P83sogA)). Would be good to get this in 3.7.3.

**Comment by RyanCavanaugh** [comment]
@typescript-bot cherry-pick this to release-3.7
